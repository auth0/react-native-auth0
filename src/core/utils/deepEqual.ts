export function deepEqual<T>(x: T, y: T): boolean {
  if (x === y) {
    return true;
  } else if (
    typeof x == 'object' &&
    x != null &&
    typeof y == 'object' &&
    y != null
  ) {
    if (Object.keys(x).length != Object.keys(y).length) return false;

    for (const prop in x) {
      if (Object.prototype.hasOwnProperty.call(y, prop)) {
        if (!deepEqual(x[prop], y[prop])) return false;
      } else return false;
    }
    return true;
  } else {
    return false;
  }
}
