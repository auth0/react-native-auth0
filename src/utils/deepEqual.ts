export function deepEqual(x: any, y: any) {
  if (x === y) {
    return true;
  } else if (
    typeof x == 'object' &&
    x != null &&
    typeof y == 'object' &&
    y != null
  ) {
    if (Object.keys(x).length != Object.keys(y).length) return false;

    for (var prop in x) {
      if (y.hasOwnProperty(prop)) {
        if (!deepEqual(x[prop], y[prop])) return false;
      } else return false;
    }
    return true;
  } else {
    return false;
  }
}
