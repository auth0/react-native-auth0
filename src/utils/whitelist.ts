import BaseError from './baseError';

export default class ParameterError extends BaseError {
  public expected: unknown[];
  public actual: string;
  public missing: unknown[];

  constructor(expected: unknown[], actual: string, missing: unknown[]) {
    super(
      'Missing required parameters',
      `Missing required parameters: ${JSON.stringify(missing, null, 2)}`,
    );
    this.expected = expected;
    this.actual = actual;
    this.missing = missing;
  }
}

export function apply(rules: any, values: any) {
  const {whitelist = true, parameters, aliases = {}} = rules;
  let mapped: {[key: string]: string} = {};
  let requiredKeys = Object.keys(parameters)
    .filter(key => parameters[key].required)
    .map(key => parameters[key].toName || key);
  for (let key of Object.keys(values)) {
    let value = values[key];
    let parameterKey: string = aliases[key] || key;
    let parameter = parameters[parameterKey];
    if (parameter && value) {
      mapped[parameter.toName || parameterKey] = value;
    } else if (value && !whitelist) {
      mapped[key] = value;
    }
  }
  let missing = requiredKeys.filter(key => !mapped[key]);
  if (missing.length > 0) {
    throw new ParameterError(requiredKeys, values, missing);
  }
  return mapped;
}
