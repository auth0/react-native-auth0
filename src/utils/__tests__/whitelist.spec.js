import {apply} from '../whitelist';
import {deepEqual} from '../deepEqual';
import faker from 'faker';

describe('whitelist', () => {
  const rules = {
    parameters: {
      state: {required: true},
      nonce: {required: false},
      clientId: {required: false, toName: 'client_id'},
      realm: {},
    },
    aliases: {
      connection: 'realm',
      clientID: 'clientId',
    },
  };

  it('should keep declared values', () => {
    const value = {
      state: faker.random.uuid(),
      nonce: faker.random.uuid(),
      clientId: faker.random.uuid(),
      connection: faker.random.word(),
    };
    expect(apply(rules, value)).toMatchObject({
      state: value.state,
      nonce: value.nonce,
      client_id: value.clientId,
      realm: value.connection,
    });
  });

  it('should fail if required key is not found', () => {
    const value = {state: faker.random.uuid()};
    expect(() => apply(rules, {})).toThrowErrorMatchingSnapshot();
  });

  it('should handle multiple parameters', () => {
    const value = {
      state: faker.random.uuid(),
      nonce: faker.random.uuid(),
    };
    expect(apply(rules, value)).toMatchObject(value);
  });

  it('should remove non declared keys by default', () => {
    const state = faker.random.uuid();
    const value = {
      state,
      initialState: faker.random.uuid(),
    };
    expect(apply(rules, value)).toMatchObject({state});
  });

  it('should keep non declared keys', () => {
    const state = faker.random.uuid();
    const value = {
      state,
      initialState: faker.random.uuid(),
    };
    const allowNonDeclared = {whitelist: false, ...rules};
    expect(apply(allowNonDeclared, value)).toMatchObject(value);
  });

  it('should consider parameters as optional by default', () => {
    const value = {
      state: faker.random.uuid(),
    };
    expect(apply(rules, value)).toMatchObject({state: value.state});
  });

  it('should use mapped name if available', () => {
    const state = faker.random.uuid();
    const clientId = faker.random.uuid();
    const value = {
      state,
      clientId,
    };
    expect(apply(rules, value)).toMatchObject({
      state,
      client_id: clientId,
    });
  });
});

describe('test deep equals', () => {
  test('returns true for equal objects', () => {
    const obj1 = {
      name: 'John',
      age: 30,
      address: {city: 'New York', state: 'NY'},
    };
    const obj2 = {
      name: 'John',
      age: 30,
      address: {city: 'New York', state: 'NY'},
    };
    expect(deepEqual(obj1, obj2)).toBe(true);
  });

  test('returns false for different objects', () => {
    const obj1 = {
      name: 'John',
      age: 30,
      address: {city: 'New York', state: 'NY'},
    };
    const obj2 = {
      name: 'Jane',
      age: 25,
      address: {city: 'Boston', state: 'MA'},
    };
    expect(deepEqual(obj1, obj2)).toBe(false);
  });

  test('returns false for objects with different nested properties', () => {
    const obj1 = {
      name: 'John',
      age: 30,
      address: {city: 'New York', state: 'NY'},
    };
    const obj2 = {
      name: 'John',
      age: 30,
      address: {city: 'Boston', state: 'MA'},
    };
    expect(deepEqual(obj1, obj2)).toBe(false);
  });

  test('returns false for objects with different numbers of properties', () => {
    const obj1 = {name: 'John', age: 30};
    const obj2 = {
      name: 'John',
      age: 30,
      address: {city: 'New York', state: 'NY'},
    };
    expect(deepEqual(obj1, obj2)).toBe(false);
  });

  test('returns false for objects with null values', () => {
    const obj1 = {name: 'John', age: 30, address: null};
    const obj2 = {
      name: 'John',
      age: 30,
      address: {city: 'New York', state: 'NY'},
    };
    expect(deepEqual(obj1, obj2)).toBe(false);
  });
});
