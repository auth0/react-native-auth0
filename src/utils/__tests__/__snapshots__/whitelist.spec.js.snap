// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whitelist should fail if required key is not found 1`] = `
"Missing required parameters: [
  "state"
]"
`;
